### PR TITLE
Add deisa plugin

### PR DESCRIPTION
**Pull request by benoitmartin88**
# Short description of the MR

Add Deisa plugin to enable in-situ analysis.

# List of things to check before making a PR

Before merging your code, please check the following:

* [x] you have added a line describing your changes to the Changelog;
* [ ] you have added unit tests for any new or improved feature;
* [x] In case you updated dependencies, you have checked pdi/docs/CheckList.md
* you have checked your code format:
  - [x] you have checked that you respect all conventions specified in CONTRIBUTING.md;
  - [x] you have checked that the indentation and formatting conforms to the `.clang-format`;
  - [x] you have documented with doxygen any new or changed function / class;
* you have correctly updated the copyright headers:
  - [x] your institution is in the copyright header of every file you (substantially) modified;
  - [x] you have checked that the end-year of the copyright there is the current one;
* you have updated the AUTHORS file:
  - [x] you have added yourself to the AUTHORS file;
  - [x] if this is a new contribution, you have added it to the AUTHORS file;
* you have added everything to the user documentation:
  - [x] any new CMake configuration option;
  - [x] any change in the yaml config;
  - [x] any change to the public or plugin API;
  - [x] any other new or changed user-facing feature;
  - [x] any change to the dependencies;
* you have correctly linked your MR to one or more issues:
  - [x] your MR solves an identified issue;
  - [ ] your commit contain the `Fix #issue` keyword to autoclose the issue when merged.

### CODE REVIEW

**Comment by benoitmartin88** [comment]
For the sake of merging this feature quickly, I will open a new issue for the tests.
This plugin has been used with success, so I feel confident that it works well.

File: plugins/deisa/deisa.cxx [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+		if (!PC_status(init_tree)) {
+			ctx.callbacks().add_event_callback([&](const std::string&) { init_deisa(); }, to_string(init_tree));
+		} else {
+			throw Config_error{conf, "Deisa plugin requires init_on key "};
+		}
+
+		// map_in
+		PC_tree_t map_tree = PC_get(conf, ".map_in");
+		if (!PC_status(map_tree)) {
+			each(map_tree, [&](PC_tree_t key_map, PC_tree_t value_map) {
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:108]: Once https://github.com/pdidev/pdi/issues/481 is fixed, we should use the `or_none` variant here.

Comment (jbigot) [inline plugins/deisa/deisa.cxx:108]:
```suggestion
		PC_tree_t map_in = PC_get(conf, ".map_in");
		if (!PC_status(map_in)) {
			each(map_in, [&](PC_tree_t key_map, PC_tree_t value_map) {
```

File: plugins/deisa/deisa.cxx [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+							pydict pyscope = pymod::import("__main__").attr("__dict__");
+							pyscope[deisa_array_name.c_str()] = to_python(data_ref);
+							pyscope["time_step"] = m_time_step.to_long(ctx);
+							pyscope["name"] = deisa_array_name.c_str();
+
+#ifdef NDEBUG
+							py::exec("bridge.publish_data(" + deisa_array_name + ", name, time_step, debug=False)", pyscope);
+#else
+							py::exec("bridge.publish_data(" + deisa_array_name + ", name, time_step, debug=True)", pyscope);
+#endif
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:122]: Maybe add an issue to make debug a configuration option in the yaml file

File: plugins/deisa/deisa.cxx [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+							pyscope[deisa_array_name.c_str()] = NULL;
+						} catch (const std::exception& e) {
+							ctx.logger().error("While publishing data, caught exception {}", e.what());
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:125]:
```suggestion
```

File: plugins/deisa/deisa.cxx [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+							throw Plugin_error("While publishing data. Caught exception: {}", std::string(e.what()));
+						} catch (...) {
+							ctx.logger().error("While publishing data.");
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:128]:
```suggestion
```

File: plugins/deisa/deisa.cxx [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+			darr["timedim"] = timedim;
+			darrs[deisa_array_name] = darr;
+			darrs_dtype[deisa_array_name] = datatype_to_pydtype(std::dynamic_pointer_cast<const Scalar_datatype>(type_sptr));
+		}
+
+		try {
+			// setup python context and instantiate bridge
+			pydict pyscope = pymod::import("__main__").attr("__dict__");
+			pyscope["deisa"] = pymod::import("deisa");
+			pymod deisa = pymod::import("deisa");
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:204]: not sure about the double import. Also this could reuse a class variable.

File: plugins/deisa/deisa.cxx [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+			pyscope["get_bridge_instance"] = deisa.attr("get_bridge_instance");
+			pyscope["scheduler_info"] = to_python(m_scheduler_info.to_ref(context()));
+			pyscope["rank"] = to_python(m_rank.to_ref(context()));
+			pyscope["size"] = to_python(m_size.to_ref(context()));
+			pyscope["deisa_arrays"] = darrs;
+			pyscope["deisa_arrays_dtype"] = darrs_dtype;
+
+			// TODO: use_ucx
+			py::exec("bridge = get_bridge_instance(scheduler_info, rank, size, deisa_arrays, deisa_arrays_dtype)", pyscope);
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:213]: I'd use class variables rather than python global variables so as not to pollute the global namespace.
```suggestion
			m_deisa = pymod::import("deisa");
			py::object get_bridge_instance = m_deisa.attr("get_bridge_instance");
			
			// TODO: use_ucx
			py::object bridge = get_bridge_instance(
				to_python(m_scheduler_info.to_ref(context())),
				to_python(m_rank.to_ref(context())),
				to_python(m_size.to_ref(context())),
				darrs,
				darrs_dtype);
```

**Review by jbigot — COMMENTED** [review]
A few propositions

File: plugins/deisa/cmake/FindPython3Path.cmake [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+################################################################################
+
+cmake_minimum_required(VERSION 3.5)
```
Comment (Yushan-Wang) [inline plugins/deisa/cmake/FindPython3Path.cmake:26]: I think 3.5 is a bit old.

Comment (Yushan-Wang) [inline plugins/deisa/cmake/FindPython3Path.cmake:26]: 3.16 as the other CMake?

**Review by Yushan-Wang — APPROVED** [review]

File: plugins/deisa/deisa.cxx [hunk 8]
```diff
@@ -98,7 +98,7 @@ class deisa_plugin: public Plugin
 			ctx.callbacks().add_event_callback([&](const std::string&) { init_deisa(); }, to_string(init_tree));
 		} else {
 			throw Config_error{conf, "Deisa plugin requires init_on key "};
-		}
+		} // TODO: replace with try/catch when #480 is fixed.
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:101]: Maybe add a comment in the issue so that we remember about it

File: plugins/deisa/deisa.cxx [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+	}
+
+private:
+	/**
+	 * Check that the PDI plugin is compatible with Deisa's python Bridge (i.e, that the python API is what it should be).
+	 * The plugin is compatible if PYTHON_LIBRARY_COMPATIBILITY == deisa.__version__.__version__ (python)
+	 */
+	static void check_compatibility()
+	{
+		pymod deisa = pymod::import("deisa.__version__");
```
[the remark this answers is not in the document]
Comment (benoitmartin88) [reply plugins/deisa/deisa.cxx]: I'm not going to, because it creates a segfault when calling `py::finalize_interpreter();` and this is only called once. If a scenario occurs where the module is used more than once, I'll consider making this a class variable.

[the remark this answers is not in the document]
Comment (jbigot) [reply plugins/deisa/deisa.cxx]: yes, it should be set to Null in the dtor before finalizing, but indeed, now that you store m_bridge, it makes no sense to store the module.

**Review by Yushan-Wang — APPROVED** [review]

File: example/example.c [hunk 10]
```diff
@@ -209,6 +210,7 @@ int main(int argc, char* argv[])
 	PDI_expose("dsize", dsize, PDI_OUT);
 	PDI_expose("psize", psize, PDI_OUT);
 	PDI_expose("pcoord", pcoord, PDI_OUT);
+	PDI_event("init");
```
Comment (jbigot) [inline example/example.c:213]: why not use event "main_loop" instead in Deisa. Moving the event might have some unexpected side issues

File: example/example.c [hunk 11]
```diff
@@ -173,7 +173,8 @@ int main(int argc, char* argv[])
 
 	PDI_expose("mpi_rank", &pcoord_1d, PDI_OUT);
 	PDI_expose("mpi_size", &psize_1d, PDI_OUT);
-	PDI_event("init");
+	int ii = 0;
+	PDI_expose("iter", &ii, PDI_OUT);
```
Comment (jbigot) [inline example/example.c:177]: Why is this required?

File: plugins/deisa/deisa.cxx [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+ */
+class deisa_plugin: public Plugin
+{
+	static constexpr char DEISA_COMPATIBLE_VERSION[] = "0.3.2"; // Used to check compatibility with Deisa's python library
+
+	bool m_interpreter_initialized_in_plugin = false; // Determine if python interpreter is initialized by the plugin
+	Expression m_scheduler_info;
+	std::unordered_map<std::string, Datatype_template_sptr> m_deisa_arrays;
+	long m_rank;
+	long m_size;
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:54]: If I'm correct, these are only used from the ctor. Not sure it makes sense to keep them

File: plugins/deisa/deisa.cxx [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+		}
+	}
+
+	~deisa_plugin() noexcept override
+	{
+		try {
+			if (m_interpreter_initialized_in_plugin) {
+				if (hasattr(m_bridge, "release")) {
+					// call bridge release() so that we can clear things before destructor is called (if it is ever called !).
+					m_bridge.release();
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:135]: please set it to null afterward, so that the garbage collector has a chance to run.

Also, this can be called whether we created the interpreter or not, no?

File: plugins/deisa/deisa.cxx [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+				);
+			});
+		}
+	}
+
+	~deisa_plugin() noexcept override
+	{
+		try {
+			if (m_interpreter_initialized_in_plugin) {
+				if (hasattr(m_bridge, "release")) {
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:133]: why this test? No error otherwise?

**Review by jbigot — COMMENTED** [review]
Just a few remaining very minor remarks

File: example/example.c (hunk 11 above)
Comment (benoitmartin88) [reply example/example.c:177]: I need to expose `iter` before the init event is triggered or there's an error: `[MPI 000001][PDI] *** error: Error while triggering event `main_loop': while referencing `iter': Right_error: Unable to grant access for value reference`

File: example/example.c (hunk 10 above)
Comment (benoitmartin88) [reply example/example.c:213]: you are right. I can use `main_loop`.

File: plugins/deisa/deisa.cxx (hunk 14 above)
Comment (benoitmartin88) [reply plugins/deisa/deisa.cxx:133]: Both of these checks can now be removed.

File: plugins/deisa/deisa.cxx (hunk 13 above)
Comment (benoitmartin88) [reply plugins/deisa/deisa.cxx:135]: yes `release` can be called anytime if the interpreter is still alive.

File: plugins/deisa/deisa.cxx [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+	 * The plugin is compatible if DEISA_COMPATIBLE_VERSION == deisa.__version__.__version__ (python)
+	 */
+	void check_compatibility()
+	{
+		py::module deisa = py::module::import("deisa.__version__");
+		const auto python_library_version = py::str(deisa.attr("__version__")).cast<std::string>();
+		if (python_library_version != DEISA_COMPATIBLE_VERSION) {
+			throw Plugin_error(
+				"Deisa PDI plugin is expecting Deisa python version " + std::string(DEISA_COMPATIBLE_VERSION) + " but found version "
+				+ python_library_version
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:145]: It would be better to use the fmt support here:
```suggestion
				"Deisa PDI plugin is expecting Deisa python version {} but found version {}", DEISA_COMPATIBLE_VERSION, python_library_version
```

File: plugins/deisa/deisa.cxx [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+			long m_size = static_cast<long>(mpi_size);
+
+			// instantiate bridge
+			py::module m_deisa = py::module::import("deisa");
+			py::object get_bridge_instance = m_deisa.attr("get_bridge_instance");
+
+			// TODO: use_ucx
+			m_bridge = get_bridge_instance(to_python(m_scheduler_info.to_ref(context())), m_rank, m_size, darrs, darrs_dtype);
+		} catch (const std::exception& e) {
+			throw Plugin_error("Could not initialize Deisa plugin. Caught exception: {}", std::string(e.what()));
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:192]: Don't think you need an intermediate here. Can you try:
```suggestion
			throw Plugin_error("Could not initialize Deisa plugin. Caught exception: {}", e.what());
```

File: example/example.c (hunk 11 above)
Comment (jbigot) [reply example/example.c]: indeed, in the `.yml`, you set 
```yaml
deisa_arrays:
  global_t:
    start: [ $iter, '($dsize[0]-2)*$pcoord[0]', '($dsize[1]-2)*$pcoord[1]' ]
```
so then, when this is evaluated at initalization, on the `main_loop` event, on line 161:
```C++
Datatype_sptr type_sptr = type_tpl->evaluate(context());
```
`iter` is evaluated.

The issue, I believe is that the the `$timedim` element of the `start` array is not used in python at this point. So this could be ignored, and the evaluation is pointless, but it is still evaluated.

Can you add an issue about supporting partial evaluation of types, documenting this use-case and then, I guess you will indeed have to expose this just before the `main_loop` event.

File: example/README.md [hunk 17]
```diff
@@ -117,3 +117,9 @@ Above instruction will share `iter` and `main_field`, call `newiter` event and t
 This is the place when plugins will read/write our data.
 
 We have covered the logic behind the %PDI example. Now you can start the \ref Hands_on.
+
+
+## Dask enabled insitu analytics (Deisa)
+
+Requirements: 
+
```
Comment (jbigot) [inline example/README.md:125]: Looks like a half done addition. Maybe just add an issue to remind ourselves to add it later for now
```suggestion
```

File: plugins/deisa/deisa.cxx [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+		try {
+			int mpi_size;
+			long m_rank = Ref_r{ctx.desc("MPI_COMM_WORLD_rank").ref()}.scalar_value<long>();
+			MPI_Comm comm = *static_cast<const MPI_Comm*>(Ref_r{ctx.desc("MPI_COMM_WORLD").ref()}.get());
+			MPI_Comm_size(comm, &mpi_size);
+			long m_size = static_cast<long>(mpi_size);
+
+			// instantiate bridge
+			py::module m_deisa = py::module::import("deisa");
+			py::object get_bridge_instance = m_deisa.attr("get_bridge_instance");
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:188]: Let's reserve `m_` for class members
```suggestion
			py::module deisa = py::module::import("deisa");
			py::object get_bridge_instance = deisa.attr("get_bridge_instance");
```

File: plugins/deisa/deisa.cxx [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+
+			// TODO: use_ucx
+			m_bridge = get_bridge_instance(to_python(m_scheduler_info.to_ref(context())), m_rank, m_size, darrs, darrs_dtype);
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:191]: Let's reserve `m_` for class members
```suggestion
			long rank = Ref_r{ctx.desc("MPI_COMM_WORLD_rank").ref()}.scalar_value<long>();
			MPI_Comm comm = *static_cast<const MPI_Comm*>(Ref_r{ctx.desc("MPI_COMM_WORLD").ref()}.get());
			int mpi_size;
			MPI_Comm_size(comm, &mpi_size);

			// instantiate bridge
			py::module deisa = py::module::import("deisa");
			m_bridge = deisa.attr("get_bridge_instance")(
				to_python(m_scheduler_info.to_ref(context())),
				rank,
				static_cast<long>(mpi_size),
				darrs,
				darrs_dtype
			);
```

File: plugins/deisa/deisa.cxx [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+using namespace py::literals;
+
+/** The deisa plugin
+ */
+class deisa_plugin: public Plugin
+{
+	static constexpr char DEISA_COMPATIBLE_VERSION[] = "0.3.2"; // Used to check compatibility with Deisa's python library
+
+	Expression m_scheduler_info;
+	std::unordered_map<std::string, Datatype_template_sptr> m_deisa_arrays;
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:51]: Could you make these two into temporary variables that you pass as a parameter to `init_deisa`

File: plugins/deisa/deisa.cxx [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+		if (python_library_version != DEISA_COMPATIBLE_VERSION) {
+			throw Plugin_error(
+				"Deisa PDI plugin is expecting Deisa python version {} but found version {}",
+				DEISA_COMPATIBLE_VERSION,
+				python_library_version
+			);
+		}
+	}
+
+	void init_deisa(Context& ctx)
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:151]: Do you think this could be made into a static function that returns the bridge?

**Review by jbigot — DISMISSED** [review]
I believe this is the last batch of comments. This ended up being a much bigger cleanup than I anticipated, but I believe this will be good in the long term

File: example/deisa_client.py [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+arrays = deisa.get_deisa_arrays()
+
+# Select data
+gt = arrays["global_t"][...]
+
+# Check contract
+arrays.check_contract()
+
+
+##########################
```
Comment (Yushan-Wang) [inline example/deisa_client.py:19]: Shall we add a small analysis example here?

File: plugins/deisa/deisa.cxx (hunk 20 above)
Comment (jbigot) [reply plugins/deisa/deisa.cxx:51]: My bad, `init_deisa` is called in a callback

File: plugins/deisa/deisa.cxx (hunk 21 above)
Comment (jbigot) [reply plugins/deisa/deisa.cxx:151]: my bad, this is used in a callback

File: plugins/deisa/deisa.cxx [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+							publish_data(to_python(data_ref), deisa_array_name.c_str(), m_time_step.to_long(ctx), "debug"_a = false);
+#else
+							publish_data(to_python(data_ref), deisa_array_name.c_str(), m_time_step.to_long(ctx), "debug"_a=true);
+#endif
+						} catch (const std::exception& e) {
+							throw Plugin_error("While publishing data. Caught exception: {}", std::string(e.what()));
+						} catch (...) {
+							throw Plugin_error("While publishing data.");
+						}
+					},
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:111]: What would you say about moving this to a dedicated `on_data(const std::string&, const Ref& data_ref)` member function?

File: plugins/deisa/deisa.cxx (hunk 21 above)
Comment (jbigot) [inline plugins/deisa/deisa.cxx:151]: what would you say about renaming this function to `on_init_event`, I feel it would make the fact that its a callback clearer

File: example/deisa_client.py (hunk 22 above)
Comment (jbigot) [reply example/deisa_client.py:19]: I believe it's a good idea, and could deserve its dedicated issue & PR

File: plugins/deisa/deisa.cxx [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+
+using namespace PDI;
+using namespace py::literals;
+
+/** The deisa plugin
+ */
+class deisa_plugin: public Plugin
+{
+	static constexpr char DEISA_COMPATIBLE_VERSION[] = "0.3.2"; // Used to check compatibility with Deisa's python library
+
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:49]: Sorry, I believe I miscommunicated and was the reason you removed this.
```suggestion

	bool m_interpreter_initialized_in_plugin = false; // Determine if python interpreter is initialized by the plugin
```

File: plugins/deisa/deisa.cxx [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+	py::object m_bridge = py::none();
+
+public:
+	static std::pair<std::unordered_set<std::string>, std::unordered_set<std::string>> dependencies() { return {{"mpi"}, {"mpi"}}; }
+
+	deisa_plugin(Context& ctx, PC_tree_t conf)
+		: Plugin{ctx}
+	{
+		if (!Py_IsInitialized()) {
+			py::initialize_interpreter();
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:62]:
```suggestion
			py::initialize_interpreter();
			m_interpreter_initialized_in_plugin = true;
```

File: plugins/deisa/deisa.cxx [hunk 26]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	~deisa_plugin() noexcept override
+	{
+		try {
+			// call bridge release() so that we can clear things before destructor is called (if it is ever called !).
+			assert(hasattr(m_bridge, "release"));
+			m_bridge.release();
+			m_bridge = py::none();
+			py::finalize_interpreter();
```
Comment (jbigot) [inline plugins/deisa/deisa.cxx:114]:
```suggestion
			if(m_interpreter_initialized_in_plugin) {
				py::finalize_interpreter();
			}
```

**Review by jbigot — APPROVED** [review]
Sorry, I missed a point. But apart from these 3 small changes that I forgot, I'd say it's good to ship.

Great cleanup work, I really believe the plugin is now in a much better situation and I'm now confident for its future!

File: plugins/deisa/deisa.cxx (hunk 24 above)
Comment (benoitmartin88) [reply plugins/deisa/deisa.cxx:49]: I don't understand. I removed this check in a previous review. You suggest that we put it back ?

Comment (jbigot) [reply plugins/deisa/deisa.cxx:49]: Yes, I miscommunicated. My point was not that this test is useless. Rather, my point was that the destruction of the bridge should be done in all cases. It used to be done only if `m_interpreter_initialized_in_plugin` was true

**Review by Yushan-Wang — APPROVED** [review]
Great!

**Review by jbigot — APPROVED** [review]
Congrats, we're there now :tada:
